component.js

```diff
@@ -3,15 +3,19 @@ var DefineMap = require("can-define/map/map");
 var assign = require("can-util/js/assign/assign");
 var Observer = require("./observer");
 var makeEnumerable = require("./make-enumerable");
+var bindMethods = require("./bind-methods");
 var dev = require("can-util/js/dev/dev");
 var namespace = require("can-namespace");
 
 if (React) {
 	var Component = function Component() {
 		React.Component.call(this);
 
-		if (this.constructor.ViewModel && !makeEnumerable.isEnumerable(this.constructor.ViewModel)) {
-			makeEnumerable(this.constructor.ViewModel, true);
+		if (this.constructor.ViewModel) {
+			bindMethods(this.constructor.ViewModel, true);
```

Comment:
Could you make it recursive and add a way to make sure methods don't become double bound? `makeEnumerable` does both of these if you need a reference.

Reply:
Make it recursive in the sense that it goes through the properties and for any maps, binds those methods to those maps? I am not sure that is necessary?

In fact we probably don't want to do it for maps we "don't control" like models and such, we just want it to be the viewModel for this component that does this binding.

What are your thought there?

Of course the double bind thing would only happen if it was recursive too right?

Reply:
Double binding would happen more frequently with recursion, but could also happen if the same VM were used for multiple Components.

Regarding recursion: not doing it for those inner maps would leave us in the same position. You couldn't have `<div onClick={ this.viewModel.counter.increment } >` (where `counter` is an instance or your Counter model).

What if, instead of recursing, there was a helper. Instead of defining `counter: { Type: Counter }`, you could define `counter: { Type: autobind(Counter) }`. That would create a new DefineMap that works just like `Counter`, but has methods autobound? That same helper would also be able to be used as a decorator, incidentally. As long as the argument to that resulting constructor can take a POJO or instance of Counter, I think it would work, though it's a bit more effort. I also don't think this would affect performance much at all (a tiny bit at initialization of a new value for `counter`).

Reply:
`makeEnumerable` is applied recursively, but I suppose that doesn't actually change any visible functionality, whereas `bindMethods` does.

Reply:
I don't know, I think of this as a convinice similar to the way `React.createClass()` would autobind methods.

Technically we could go the route of making people do all this binding in the ViewModels `init()` method, but it feels a little messy of fidly, but honestly so does having to import and then use a decorator/utility function. 

I can't think of a good reason *not* to autobind vm methods. If it wasn't a problem with `createClass()` it's probably not a real problem.

But I can see an issue with recursive autobinding, like you said it changes the behaviour of things outside this components influence.

With the `<div onClick={ this.viewModel.counter.increment } >` example you could solve that in the viewModel with a method like:
```js
increment() {
  this.counter.increment();
}
/*...*/
<div onClick={ this.viewModel.increment } >
```

...and so I think as long as the actual viewModel instance as it's methods bound, it would be sufficent.


The big problem we're really trying to solve here is the 80% easy, where instead this:
```jsx
const AppComponent = reactViewModel(ViewModel, (viewModel) => {
  return (
    <div onClick={ () => viewModel.increment() }>
      Count: {viewModel.count} (Click Me)
    </div>
  );
});
```
it can just be
```jsx
const AppComponent = reactViewModel(ViewModel, (viewModel) => {
  return (
    <div onClick={ viewModel.increment }>
      Count: {viewModel.count} (Click Me)
    </div>
  );
});
```

Reply:
After a quick video chat we decided the `autoBindMethods()` decorator will get it's own issue and I'll add a check to make sure we don't run the autoBind setup twice for the same DefineMap class